sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleServer.scala

```diff
@@ -366,74 +356,71 @@ class RapidsShuffleServer(transport: RapidsShuffleTransport,
-              serverConnection.send(
-                transferRequest.executorId,
-                AddressLengthTag.from(transferResponse.acquire(), transferRequest.responseTag()),
-                transferResponseTx => {
-                  transferResponse.close()
-                  transferResponseTx.close()
-                })
-
-              // wake up the bssExec since bounce buffers became available
-              logDebug(s"Buffer send state ${buffersToSend.tag} is done. Closing. " +
-                  s"Still pending: ${pendingTransfersQueue.size}.")
-              bssExec.synchronized {
-                bufferSendState.close()
-                bssExec.notifyAll()
+      serverStream.sync()
+
+      // need to release at this point, we do this after the sync so
+      // we are sure we actually copied everything to the bounce buffer
+      bufferSendStates.foreach(_.releaseAcquiredToCatalog())
+
+      bssBuffers.foreach {
+        case (bufferSendState, buffersToSend) =>
+          val transferRequest = bufferSendState.getTransferRequest
+          serverConnection.send(transferRequest.executorId(), buffersToSend, bufferTx =>
+            try {
+              logDebug(s"Done with the send for ${bufferSendState} with ${buffersToSend}")
+
+              if (bufferSendState.hasNext) {
+                // continue issuing sends.
+                logDebug(s"Buffer send state ${bufferSendState} is NOT done. " +
+                    s"Still pending: ${pendingTransfersQueue.size}.")
+                bssExec.synchronized {
+                  bssContinueQueue.add(bufferSendState)
+                  bssExec.notifyAll()
+                }
+              } else {
+                val transferResponse = bufferSendState.getTransferResponse()
+
+                logDebug(s"Handling transfer request for ${transferRequest.executorId()} " +
+                    s"with ${buffersToSend}")
+
+                // send the transfer response
+                serverConnection.send(
+                  transferRequest.executorId,
+                  AddressLengthTag.from(transferResponse.acquire(), transferRequest.responseTag()),
+                  transferResponseTx => {
+                    transferResponse.close()
+                    transferResponseTx.close()
+                  })
+
+                // wake up the bssExec since bounce buffers became available
+                logDebug(s"Buffer send state ${buffersToSend.tag} is done. Closing. " +
+                    s"Still pending: ${pendingTransfersQueue.size}.")
+                bssExec.synchronized {
+                  bufferSendState.close()
+                  bssExec.notifyAll()
+                }
               }
-            }
-          } finally {
-            bufferTx.close()
-          }
-        }
-      })
+            } finally {
+              bufferTx.close()
+            })
+      }
+    } catch {
+      case t: Throwable =>
+        bufferSendStates.safeClose(t)
```

Comment:
Can probably use `closeOnExcept(bufferSendStates)` for the `try` block rather than needing an explicit `catch` here.